### ISSUE TITLE
linux/arm64/v8 Architecture images

### ISSUE THREAD

**Issue by chris-cheshire** [BioDocker/containers#425]
Hi, 

We are starting to be issued with M1 macs at our institution and we are getting docker errors like:

```
docker: Error response from daemon: image with reference biocontainers/biocontainers:v1.2.0_cv1 was found but does not match the specified platform: wanted linux/arm64/v8, actual: linux/amd64.
```

I guess this is because either the program is not available as an arm compiled executable, or that biocontainers doesn't build against arm as part of its build process.

Could anyone comment on this?

**Comment by chris-cheshire** [comment]
@osallou any comments on this? Happy to help implement but its quite a large change.

**Comment by osallou** [comment]
hi,
we do not build (yet) for other archs than amd64.
There is a plan to do so (at least arm), but for that we need to get some related hardware. We expect to have one arm server next year and add container builds for arm.

Be also aware that some tools may not build on arm....

**Comment by chris-cheshire** [comment]
Hi @osallou hows the state of ARM looking? You probably know that Apple have stopped selling intel macs and my laptop is on the way out so its becoming critical for everyone at the Francis Crick that this is sorted. Is there a way I can get involved and help move it forward?

**Comment by osallou** [comment]
Hi,
we have no arm server unfortunately, and none planned for the moment...
We may benefit arm servers from Amazon in future, but nothing sure for the moment. (@bgruening  ?)

**Comment by bgruening** [comment]
Yes, the plan is to use some AWS infrastructure.

**Comment by chris-cheshire** [comment]
Ok I think this is going to become a real issue very soon. So many people test their nextflow pipelines locally on macos using docker - on an M1 this is currently impossible. Is there any timeline at all for this, is it a funding issue or development issue etc?

**Comment by bgruening** [comment]
It's more a priority and time issue. If the nextflow community needs it, they are welcome to contribute. E.g. adding support to bioconda-utils etc ...

**Comment by osallou** [comment]
Won t fix issue of arm based containers, need servers anyway

**Comment by chris-cheshire** [comment]
Yes I am very happy to contribute if that will help and I am sure I can get a few others as well if you need man-power. @osallou are you saying that AWS provide your server infrastructure for free but that they currently don't allow you to spin up ARM instances?

**Comment by osallou** [comment]
Aws will provide us compute resources among which we should have arm, but this is not yet the case

**Comment by chris-cheshire** [comment]
OK is there a timeline for when AWS will provide these?

**Comment by CharlotteAnne** [comment]
I'd just like to second the urgent need for this - use of MacOS is incredibly common in the scientific community. If there's anything I can do to help please let me know.

**Comment by osallou** [comment]
rebuilding everything for amr will take time anyway...  but our main issue is to have an arm build server

**Comment by pditommaso** [comment]
I don't know much biocontainers build infra, but docker build can target both arm and amd architecture irrespective of the host chipset via emulator. Would that not work for biocontainers building as well?

**Comment by osallou** [comment]
Seems possible via buildx but on docker desktop only (windows/mac, and not free)

We build with docker ce, on linux

**Comment by pditommaso** [comment]
There are plenty choices nowadays to build containers such as podman, buildah, kaniko, etc but not sure about the multiarch support of these. I'll made some research about this.

**Comment by pditommaso** [comment]
There's a [nice post showing](https://danmanners.com/posts/2022-01-buildah-multi-arch/) how to use [Buildah](https://github.com/containers/buildah) to build and push a multi-arch container images. 

Also considering the changes in Docker license and the OSS nature of Biocontainers, think should consider moving away from Docker for container building.

**Comment by bgruening** [comment]
It has nothing to do with Docker. We are not building Containers here, we are building packages with conda. I think what we need is needed the following:

* get ARM/RISC support into bioconda-utils (take inspiration from conda-forge)
* implement an allow-list to be consumed by bioconda-utils where we add packages that should be compiled for ARM and/or RISC
* figure out against which ARM architecture we actually want to build - which is most compatible
* getting some computational build infrastructure in place to actually compile it

We have been working so far on the last item and a grant for the first three has not been granted in 2020. If anyone wants to help you are all very welcome.

**Comment by pditommaso** [comment]
I see, out of curiosity what's your build infra now? why something like GitHub action does not fit your needs?

**Comment by bgruening** [comment]
We can use whatever public CI system that is available. But it needs to support the correct (point 3 above) ARM architecture, ideally RISC as well - and all of that in a substantial amount.

Xref:
https://github.community/t/ask-about-github-hosted-runners-arm64-platform-support/167483

**Comment by bgruening** [comment]
All of that said, I would like to manage expectations here ... this is nothing that we can do in a few weeks. (We are already trying this for a few years). And there are larger discussions to be taken in the overall scientific community. 

* It's ecological and financial probably a bad idea to create ARM containers for __all__ packages.
* The promises that we give to easily scale up a reproducible pipeline will not hold true anymore - at least to this degree. You very likely changing from M1/M2 to x86-64 based execution environments and with this, you are changing packages/containers.

There are good reasons we want ARM/RISC containers and large ARM HP clusters is one of them.
OSX users can and maybe should for the foreseeable future use a VM. This is possible isn't it?

**Comment by osallou** [comment]
On mac, i think docker uses a microvm in fact, like on windows (but not 100% sure)

**Comment by osallou** [comment]
And even for non conda packages, other ones needs to be compiled on arm, which may not work for all tools, on a base distro arm compatible too with all their dependencies.

Building a tool made for x86 is not fully transparent for all tools

**Comment by pditommaso** [comment]
So, think we are discussing two topics here: 

1. addressing the issue for Apple Silicon users, I don't think there's the need to rebuild Bioconda containers for that. As mentioned before, Docker made excellent work in this regard, it's just enough to use the CLI option `--platform linux/amd64` or the env variable `DOCKER_DEFAULT_PLATFORM=linux/amd64` to run any existing container on Mac M1 (with rare exceptions). It would be interesting to learn from @chris-cheshire if he thinks is a viable solution, and why not otherwise. 

2. the building of Bionconda/Biocontainers targeting ARM architecture for production workload and optimization reasons. I argue there could be a strategic need to support all recipes, but maybe deserves a separate thread.

**Comment by chris-cheshire** [comment]
Thanks for everyones input! its great to get some dialogue going around this. @pditommaso I will do some testing around this running nf-core pipelines in test mode on an M1 mac this week and get back you all.

**Comment by chris-cheshire** [comment]
We have been doing some testing and it appears Docker Desktop has been upgraded as with the latest version the biocontainer images seem to run by default on nf-core pipelines without any extra config. We need to test some more pipelines but this is very promising!

**Comment by edmundmiller** [comment]
@delagoya Following up on this post Nextflow Summit. Thoughts on getting together a hackathon for a day or two to get started on the effort?

cc @drpatelh @maxulysse @ewels

**Comment by chris-cheshire** [comment]
I would be interested in participating in this

**Comment by delagoya** [comment]
@Emiller88 absolutely! I need to get over the next few weeks of conferences, so earliest I will be coordinating activities is mid-December

**Comment by brainstorm** [comment]
Hello everybody in this issue :)

Some folks have recently started a new push for Linux/Aarch64 for [many Nextflow-friendly pipelines](https://github.com/ewels/nf-core-arm-discovery) and I'm myself [going through the process of migrating one of them mainly through Bioconda](https://github.com/nf-core/oncoanalyser/pull/134)... over nf-core Slack a few of us were wondering what's the status of this issue:

1. There are CI/CD workers already packing Bioconda packages into Biocontainers.
2. It's being explored by `<insert somebody's name here>`.
3. There are no plans and this issue is stale since last message 3 years ago.

Which one of the above is true(r)?

**Comment by bgruening** [comment]
1) :)

**Comment by mboudet** [comment]
As far as I know, on the bioconda side arm packages have been built since 2024/03 (https://bioconda.github.io/developer/aarch64.html)

On the 'biocontainer' side (this repo), we try to build arm containers too (since last july), but it's more of a 'best-effort' build. (And mostly work on 'simple' containers)

**Comment by brainstorm** [comment]
Fantastic, thank you!

**Comment by dslarm** [comment]
Is there somewhere to find a list of the containers built now?   There's been a _lot_ of aarch64 enablement in the last few weeks in bioconda.

However, if I go to (eg) bwa-mem2's biocontainer by searching: https://biocontainers.pro/registry?all_fields_search=bwa-mem2&sort_order=asc&sort_field=default&offset=0&limit=30 - then the only containers showing are amd64.  It's tracking the right builds (we updated bwa-mem2 a couple of weeks ago).

Where am I going wrong?  Is there another issue at play?

**Comment by bgruening** [comment]
You can grep in the bioconda recipes, the YAML files to find which tools are available for ARM.

Please do NOT use biocontainers.pro anymore, it's not maintained. @ypriverol can you please redirect this website to quay.io? Thanks.

**Comment by ChristianKniep** [comment]
Can you start building multi-arch images for tooling and packages?

```
$ manifest-tool inspect quay.io/biocontainers/fastp:0.24.0--heae3180_1
Name: quay.io/biocontainers/fastp:0.24.0--heae3180_1 (Type: application/vnd.docker.distribution.manifest.v2+json)
      Digest: sha256:fe647f76dbda3af58a4e159ec4662b7c11f29ee65a44a9b8f43ba018fee8e8a5
        Size: 751
          OS: linux
        Arch: amd64
$ manifest-tool inspect public.ecr.aws/biocontainers/fastp:0.24.0--heae3180_1
Name: public.ecr.aws/biocontainers/fastp:0.24.0--heae3180_1 (Type: application/vnd.docker.distribution.manifest.v2+json)
      Digest: sha256:c37387e48425fb06a2d5d1592046fcf3e365ca6c8779268499ce2d5d3f3e9862
        Size: 945
          OS: linux
        Arch: amd64
```

**Comment by dslarm** [comment]
@bgruening - did you have any opinion on the multi-arch images?   

Is this a good direction or bad direction to take this?
